### ISSUE TITLE
Please, steal ideas from bomi and expand options for outputs

### ISSUE THREAD

**Issue by v-fox** [kokoko3k/xt7-player-mpv#6]
For a while the GUI mplayer/mplayer2/mpv champion was mpv fork [bomi](https://github.com/xylosper/bomi) which constantly synchronized its mpv core with upstream and built upon it. But now its developer [has gone inactive](https://github.com/xylosper/bomi/issues/412) which leaves only Xt7 as the only GUI trying to utilise mpv's potential. And doing it the right way too.

However, bomi developer is quite successful in his work and has made quite a few unique features that everyone should envy. Like:
- support for multiple simultaneous subtitle tracks. and check [this idea](https://github.com/xylosper/bomi/issues/426) out
- relative text sizing for OSD and subtitles that scales with the size of the video output
- mind-boggling frame preview for any place on the time bar
- slick "playback information"
- coolest spectrogram ever for audio
- simplistic yet powerful menu structure and key binding options
- skins for whole control panel holding the time bar

What both Xt7 and bomi are lacking is customization of options for video and audio outputs. Latest release of Xt7 has shamelessly bare options for "opengl" output and does not allow addition of custom user options while mpv's manual is filled with tens of them.

Xt7 also likes to un-maximize itself on each fullscreen/non-fullscreen view cycle. Which I worked around by forcing its main windows to always be maximized by openbox.

Also, to avoid bomi's fate I would recommend somehow getting together some kind of a team. Strange that it didn't become some kind of official GUI for it with the help of core mpv devs.

**Comment by kokoko3k** [comment]
Pretty lenghty "issue", it will take a while to close this :)
Il'' try to answer all of the question and suggestions yuo made,

**Multiple simultaneus subtitle tracks:**
Definitely. I'll have to think a way to add them without breaking anything.

**relative text sizing for OSD and subtitles that scales with the size of the video output**
Why do you need that? Do you use to play different streams at different output resolutions?

**mind-boggling frame preview for any place on the time bar**
I don't know if is similar to what bomi does, but Xt7 already does that since long time, just check "Live preview when seeking" in GeneralOptions -> Video.
Then, just dragging the seekbar and keeping the Left mouse button pressed will produce a popup showing a slowmotion live preview of the selected position.
Releasing the button while in the seekbar area will seek; release the button outside of the seekbar will abort the seek operation.

**slick playback information**
There is already an hotkey bounded by default to "shift+backspace", configurable under the hotkey window ("show runtime infos") 

**coolest spectrogram ever for audio**
Hard to do, since xt7 talks to mpv via pipes, and mpv doesn't export audio info anymore, because they removed the export audio filter in a recent commit.
In the past i used them to visualize volume meters and realtime frequency distribution.
I hope something will change in the future.
Open a feature request to mpv devs to implement audio visualization, i'll support them.

**simplistic yet powerful menu structure and key binding options**
Actually you can "keybind" a lot of functions; have you looked at the hotkeys window?
popup-menus under the video area are not an option, because i prefer bind the right mouse button too, so that you are free to bind complex mouse input actions like "right button + wheel" or "right button +left button" and so on.
Just cdouble click on an empty cell in hotkey window to associate an input action to a command.

**skins for whole control panel holding the time bar**
Without touching too much the code, it could be possible to change the button image frames and the appearence of the seek bar.
By now you can drag a picture into the control panel and it will be set as background.
The setting is under "general options" -> style. (checkout the latest git for that)
To reset, just empty the the field and press return/enter.

**Audio Output options**
Xt7 is full of options for audio, it even has a ladspa plugin wrapper.
Do you need something else?

**Video output options**
Is true, mpv is full of options for opengl.
Actually i find the syntax a bit difficult to be translated to a gui application, and mpv devs are constantly changing it.
I'll take a look to them again 
Meanwhile, xt7 allows you to specify an output driver with options by using the special "custom video"  driver.
Just write: opengl:youroptions

**un-maximize issue**
Definitely to fix. (edit: fixed)

**avoid bomi's fate...**
I'm developing xt7 alone since Fri May 15, 2009.
it used qt3 toolkit and mplayer as a backend; never had a single person submitted a single line of code.

**official GUI**
It will probably never happens because xt7 don't use libmpv (binding problems) nor json ipc (bad for performance), and mpv devs don't like that.
Last time i asked wm4 (mpv dev) for help, even if the method i use is supported (not deprecated) he answered that way: 
https://github.com/mpv-player/mpv/commit/8d414e2fe7f79f1bbf9158e8aa034e9ac9cf2222#commitcomment-13952461

:)

**Comment by kokoko3k** [comment]
I just checked th "unmaximization issue". Seems to be openbox related, i'll need to find out why.
-edit-
fixed by https://github.com/kokoko3k/xt7-player-mpv/commit/3562ab036470205fd2ad803b9180f20129380181

**Comment by v-fox** [comment]
> relative text sizing for OSD and subtitles that scales with the size of the video output
> Why do you need that? Do you use to play different streams at different output resolutions?

Static font sizes are usually "calibrated" to specific resolution, that most likely is fullscreen resolution of user's monitor. But then they look too big in windowed mode and cannot be universal for devices of any resolution (which I would like them to be for usage in a live distribution).

> mind-boggling frame preview for any place on the time bar
> I don't know if is similar to what bomi does, but Xt7 already does that since long time, just check "Live preview when seeking" in GeneralOptions -> Video.
> Then, just dragging the seekbar and keeping the Left mouse button pressed will produce a popup showing a slowmotion live preview of the selected position.
> Releasing the button while in the seekbar area will seek; release the button outside of the seekbar will abort the seek operation.

Hah, that's quite cool indeed ! But its size doesn't seem to be configurable, unfortunately. And I mostly watch fullscreen videos from the couch while controlling with the Bluetooth DualShock 3 gamepad. For bomi I set it to about a quarter of the screen and it's "static" too so it comes up instantly for any video quality just by dragging a mouse over the time bar.

> slick playback information
> There is already an hotkey bounded by default to "shift+backspace", configurable under the hotkey window ("show runtime infos")

Mmm, and look similar too ! Quite nice. I've noticed a problem with it thought: it shows up for a limited time and not enough to look at it properly. And if I would press the hotkey several times, those commands seem to queue up and info "jumps" in and out of existence.

> coolest spectrogram ever for audio
> Hard to do, since xt7 talks to mpv via pipes, and mpv doesn't export audio info anymore, because they removed the export audio filter in a recent commit.
> In the past i used them to visualize volume meters and realtime frequency distribution.
> I hope something will change in the future.
> Open a feature request to mpv devs to implement audio visualization, i'll support them.

:( Well, that sucks. It seems that when they do so they don't even bother to check if there is actual usage for such things and what it is. 

> official GUI
> It will probably never happens because xt7 don't use libmpv (binding problems) nor json ipc (bad for performance), and mpv devs don't like that.

Like they did those interfaces that no-one uses and screw with the ones that are used. For me they threw quite a tantrum with insults about how smplayer forces "obsolete" options but this is somehow my fault.

"GUI authors should use proper interface", they say. libmpv with pure Qt4/5 looks like clean solution, I wonder why bomi's author didn't go that way. And I'm pretty sure that smplayer's author does things pretty similarly to you. Any other mpv "GUIs" are poorer to even console interface even if they do use libmpv.

> simplistic yet powerful menu structure and key binding options
> Actually you can "keybind" a lot of functions; have you looked at the hotkeys window?
> popup-menus under the video area are not an option, because i prefer bind the right mouse button too, so that you are free to bind complex mouse input actions like "right button + wheel" or "right button +left button" and so on.
> Just cdouble click on an empty cell in hotkey window to associate an input action to a command.

Yeah, I glanced at it and quantity of possible binds seemed quite scanty in comparison to bomi's. I'm not sure if it's because I missed something or bomi's author added some non-mpv ones or you missed some of them.
Missing right menu is probably no big deal, especially if control panel is rich on options and all kinds of funky binds are possible.

And about control panel, it's kinda weird how its upper part "moves" the video in fullscreen mode.

> skins for whole control panel holding the time bar
> Without touching too much the code, it could be possible to change the button image frames and the appearence of the seek bar.
> By now you can drag a picture into the control panel and it will be set as background.
> The setting is under "general options" -> style. (checkout the latest git for that)
> To reset, just empty the the field and press return/enter.

Interesting. My real issue with the control panel is low contrast for colours of filled/unfilled parts of time and volume bars. Which are probably taken from current Qt4's style. So, I will probably may workaround that. But options for their colours would be easier than fiddling with colour palettes (especially if KDE's colour palette editor is missing from the system).

> Audio Output options
> Xt7 is full of options for audio, it even has a ladspa plugin wrapper.
> Do you need something else?

Actually, for me it's fine. I use JACK output and direct it to JAMin where same ladspa plugins do compression and equalization.

> Video output options
> Is true, mpv is full of options for opengl.
> Actually i find the syntax a bit difficult to be translated to a gui application, and mpv devs are constantly changing it.
> I'll take a look to them again
> Meanwhile, xt7 allows you to specify an output driver with options by using the special "custom video" driver.
> Just write: opengl:youroptions

That's true, they fiddle with them every release. I should really look into "custom video driver".

> un-maximize issue
> Definitely to fix.
> I just checked th "unmaximization issue". Seems to be openbox related, i'll need to find out why.

You mean it happens only under openbox, huh ? Quite ironic: openbox to break and openbox to workaround that.

> avoid bomi's fate...
> I'm developing xt7 alone since Fri May 15, 2009.
> it used qt3 toolkit and mplayer as a backend; never had a single person submitted a single line of code.

I see, and that's a shame. It's like people don't use their media players with GUIs under Linux or something. Well, most use VLC but its interface sucks, the very example of cluttered dysfunctional mess and it doesn't update often too or trying to fix any of that.

**Comment by kokoko3k** [comment]
> Static font sizes are usually "calibrated" to specific resolution, that most likely is fullscreen resolution of user's monitor. But then they look too big in windowed mode and cannot be universal for devices of any resolution (which I would like them to be for usage in a live distribution).

Font sizes should depend on the display dpi, so, in theory, even in a live environment, given that the dpi is correctly set/calculated, and not set to fixed values (usually 96dpi), they sohuld look fine anytime.
My concern is that subtitles are meant to be read, and not to be "nice in window mode".
So my use case is that one set a pleasant size in the config and will be able to read that sub even if the video window is small.
What is the point of reducing the subtitle if one cannot read it because it is too small?

I'm really open to understand your point.

> For bomi I set it to about a quarter of the screen and it's "static" too so it comes up instantly for any video quality just by dragging a mouse over the time bar.

I don't plan to make it static, PictureInPicture is funny enough :)
As for the size, it is not statically encoded, but depends from the actual screen that way:
Thumbnail width = (Screen width / 6)
Thumbnail height is calculated to match the aspect ratio.
Is that too small for you?

> it shows up for a limited time and not enough to look at it properly.

I know, as per Xt7 architecture, it is not that easy to made them stick to the display.
They last about 3 seconds, every time you push the hotkey, the counter begins from 0
So push the hotkey every 2 seconds and you should see them without problems.

> And about control panel, it's kinda weird how its upper part "moves" the video in fullscreen mode.

Yes, it is, i'll look into it, but don't hold your breath :D

> My real issue with the control panel is low contrast for colours of filled/unfilled parts of time

The seekbar color changes accordingly to the qt4 palette.
Could you provide a screenshot of the low contrast thing?

> I use JACK output and direct it to JAMin

This is good only if you have an external, separate amplifier to modify the volume.
If you modify the volume with the player, then you will feed jamin with different signal levels, and his compression will act strangely.
I use fastlookaheadlimiter with ladspa to keep high the volume, and setup a chain in xt7:
ladspa fastlookaheadlimiter filter -> volume filter

as for bomi, bomi author used to incorporate the whole mpv code into bomi because mpv keeps changing things very fast.

**Comment by v-fox** [comment]
> My concern is that subtitles are meant to be read, and not to be "nice in window mode".
> So my use casse is that one set a pleasant size in the config and will be able to read that sub even if the video window is small.
> What is the point of reducing the subtitle if one cannot read it because it is too small?
> I'm really open to understand your point.

Static sizes make sense as long as you don't go to a couch when your video runs fullscreen or something. This is why I set subtitle size for fullscreen 3-5 times bigger of what I would like them to be if I would sit right in front of the screen. bomi scales them automatically, so it's not an issue.
And I'm not sure how subtitles are scaled between differently sized monitors. If they scale fine as it is then it may be enough just to provide option for setting alternative size for fullscreen.

> I don't plan to make it static, PictureInPicture is funny enough :)
> As for the size, it is not statically encoded, but depends from the actual screen that way:
> Thumbnail width = (Screen width / 6)
> Thumbnail height is calculated to match the aspect ratio.
> Is that too small for you?

Yeah, about 4-6 times smaller than I would like it to be. And even by that logic thumbnail should be about 320x180 for 1920x1080 but it looks 2 times smaller than that. That scaling should probably be an option.

> ```
> I know, as per Xt7 architecture, it is not that easy to made them stick to the display.
> They last about 3 seconds, every time you push the hotkey, the counter begins from 0
> So push the hotkey every 2 seconds and you should see them without problems.
> ```

That's weird and quite an ugly workaround. And this is exactly what I did, it started "flashing" in and out. If something in mpv prevents it from being toggled it's probably should be reported to its devs as an error.

> ```
> And about control panel, it's kinda weird how its upper part "moves" the video in fullscreen mode.
> Yes, it is, i'll look into it, but don't hold your breath :D
> ```

Maybe it would be easier to move controls from the upper bar to the lower one and remove the upper bar from fullscreen altogether ?

> ```
> The seekbar color changes accordingly to the qt4 palette.
> Could you provide a screenshot of the low contrast thing?
> ```

![20151103-220721](https://cloud.githubusercontent.com/assets/362763/10915188/81da11e8-8277-11e5-863b-03a43c9adf91.jpg)

It's really bad to watch from any distance more than 0.5m from screen.

> ```
> This is good only if you have an external, separate amplifier to modify the volume.
> If you modify the volume with the player, then you will feed jamin with different signal levels, and his compression will act strangely.
> I use fastlookaheadlimiter with ladspa to keep high the volume, and setup a chain in xt7:
> ladspa fastlookaheadlimiter filter -> volume filter
> ```

Well, it is quite good for me. I've set up constant volume on my big speakers to the maximum amount that I could tolerate and set up JAMin so it would not let volume going to the sound card be more than proper 0db but also would increase the volume if it's too much below 0db. And in ALSA mixer I've set perfect volume which pretty much is never off.
So now, volume always and everywhere is pretty good but if I would want to be noisy, I just need to adjust ALSA mixer's slider.

Of course in that setup I never change player's volume and only rely on its normalizer but even that is unnecessary.
You can check out that setup on [my live builds](https://sourceforge.net/projects/hackeurs-sans-frontieres/) or see [fresh configs on github](https://github.com/v-fox/live_opensuse_hsf/tree/master/data/common-userfiles).

> as for bomi, bomi author used to incorporate the whole mpv code into bomi because mpv keeps changing things very fast.

Yeah, probably. And on other hand, mplayer doesn't change them enough. There is no winning.

**Comment by kokoko3k** [comment]
> Static sizes make sense as long as you don't go to a couch 

Oky, you know what? I just discovered that actually the font scales accordingly to the height of the window.
I'm not sure when this started to happen, maybe a core change in mpv.
Is ok/does it scales for you?

> That scaling should probably be an option.
> /../
> It's really bad to watch from any distance more than 0.5m from screen.

Both fixed in latest git,
There is an option to scale the live preview size and to choose a different color for the seekbar.

**Comment by kokoko3k** [comment]
> And about control panel, it's kinda weird how its upper part "moves" the video in fullscreen mode.

Now fixed

**Comment by kokoko3k** [comment]
> it shows up for a limited time and not enough to look at it properly.

Fixed; press to show, press again to hide.

**Comment by kokoko3k** [comment]
So, to summarize:

**Fixed:**
- Relative text sizing for OSD and subtitles that scales with the size of the video output (it seems to scale accordingly to the height of the window (tested with vdpau,opengl and xv)
- mind-boggling frame preview for any place on the time bar (was already implemented)
- slick "playback information" (fixed the hotkey handling)
- un-maximize problem with openbox

**To do:**
- support for multiple simultaneous subtitle tracks. 
- more opengl options

**Undecided:**
- coolest spectrogram ever for audio

Needs "upstream" changes
- simplistic yet powerful menu structure and key binding options

I didn't understood if keybindings are enough for you and what are the problems with the menus.
- skins for whole control panel holding the time bar

...Are you going to provide skins? :)

**Comment by v-fox** [comment]
> So, to summarize:

Cool, I'll look into it on the release (when it will get into "Packman" openSUSE repository, to be precise) !

> more opengl options

Meanwhile, I figured out "custom video driver" option. Works great !
Even thought "interpolation" seems broken on my system and there are some issues with getting ICC profiles from EDID. But that's mpv issues.

> Undecided:
>     coolest spectrogram ever for audio

I know it's probably hard... but look at it ! It's glorious !

> I didn't understood if keybindings are enough for you and what are the problems with the menus.

I'm not sure how to put it into words even. In bomi pretty much every action is present in menus but at the same time they don't seem cluttered. It's a minor thing and having most useful options on a control panel instead of menus is probably much more convenient in the long run.

I didn't actually configured keybindings yet but at fast glance the list of all available bindings seems bare in comparison. I might be completely wrong on that.

> skins for whole control panel holding the time bar
> 
> ...Are you going to provide skins? :)

Naah. It also may be completely unnecessary since even in bomi I find only 2 skins of any use: "native" and "air". Xt7's default isn't fancy but it isn't bad either, it's quite adequate. So, it's more of a taste thing. Personally, I want icons on control panel to be colourful and distinguished from each other.
It's been a pain to get rid of all "modern" "monochrome" hipster bullshit icons from my system and I still couldn't get system-wide player control icons right.

**Comment by v-fox** [comment]
Ahh... and one more thing: why is clicking on "volume" button on the control panel hides/unhides volume bar and is there a button for mute/unmute ?)

**Comment by kokoko3k** [comment]
No button for mute/unmute; clicking hides/show the volume bar because some may like it, some may want to just use the wheel over the icon (on info are lost, because the actual volume is shown on the statusbar when you change it via scrollwheel).
Maybe middleclicking on the button could mute/unmute (?)

-edit
Yes: https://github.com/kokoko3k/xt7-player-mpv/commit/794311944a67081f75aa02f1d699459a0e93fd30
send me a beer :P

**Comment by v-fox** [comment]
> Maybe middleclicking on the button could mute/unmute (?)

I tried to press it. Doesn't seem to work that way now.

> No button for mute/unmute; clicking hides/show the volume bar because some may like it, some may want to just use the wheel over the icon (on info are lost, because the actual volume is shown on the statusbar when you change it via scrollwheel).

Technically It's true but panel space conservation is minuscule to the point of no practicality. I think it would make more sense if left click would mute/unmute and right click - hide/unhide.

Same goes for "zoom" button: it's weird how left click just repeats its tooltip help info, it would be easier if current right click functions (zoom mode cycling) were on the left click.

And since there is a dedicated zoom button maybe a speed button would go along (to the left from zoom button) with it nicely ? You know how professional sound editors have "rewind speed wheel" or something like that ? In simple form it's a bar with slider in the middle which resets its position when mouse button is un-pressed. When it is pressed the speed of playback is dynamically set according to slider's inclining (maximum value, of course, is configurable): faster for the right, slower for the left, zero/"stop" for the middle. Right button in that setup may hide the bar similarly to the volume bar and middle - reset to middle. Modes of operation may be "sticky" and "plain" - "speed is permanently set even after left button is un-press" and "reset/stop after un-press" respectively.

**Comment by kokoko3k** [comment]
The new function of mute/unmute is in the latest git, if you're using packman builds, then you'll have to wait, i guess.

About the zoom button, you know how it acts because you clicked on it with the left mouse button, i made it that way to "instruct" users, and it seems to work; i don't think it is a big deal once one have read the message.

About the speed, there are already buttons to speedup/slowdown playback with visual feedback on the video window showing the actual speed multiplier factor.
Why one would need functionalities like non linear video editing applications in a player? mpv is able to seek forward and backward instantly, i find it overcompliated/not worth the hassle.
If one really needs to control the playback speed often, then there are hotkeys for that...

**Comment by v-fox** [comment]
> About the zoom button, you know how it acts because you clicked on it with the left mouse button, i made it that way to "instruct" users, and it seems to work; i don't think it is a big deal once one have read the message.

I propose that way so buttons for "main function" (left) and "hide function" (right) would be consistent for all elements of control panel. Help pop ups and element hiding isn't really what people expect while doing left clicks. Especially if there is a hover tooltip and elements aren't really big or otherwise intrusive.

> About the speed, there are already buttons to speedup/slowdown playback with visual feedback on the video window showing the actual speed multiplier factor.

Huh, that's what those buttons do ?! So, you actually did that, hah. But slider would allow setting speed from -10x to 10x pretty much instantly while with plain buttons one has to perform Windows-style "mouse jerk-off" which isn't fun at all.

> Why one would need functionalities like non linear video editing applications in a player? mpv is able to seek forward and backward instantly, i find it overcompliated/not worth the hassle.

I would call it "abruptly" rather than "instantly". It especially bad since all modern lossy video compression algorithms rely on constant motions and do not perform compression of individual frames, even if player does show still picture it's most likely a result of calculation of data from a multiple frames.
Sometimes (more likely - most of the time) if you want to seek a specific fragment, you want to promptly skip all unrelated parts but be more precise when you near what you're looking for.

In fact, I don't quite understand why this feature is considered an "editing" one and not a "playback" one. I figure it would be more useful in a player. Only once in my life I've stumbled upon a video on which permanent speed modification would be practical but smooth seeking is something which is needed every day.

> If one really needs to control the playback speed often, then there are hotkeys for that...

Wheel is a wheel because a key can't be so smooth, fast and precise. in fact, I wouldn't mind a mouse having multiple wheels.

**Comment by kokoko3k** [comment]
It is not a feature many users want (i think), if you like wheels, bound an hotkey to (say) right mouse button + wheel = "add speed +0.01" and "add speed -0.01" ; that way you can change speed by 1%.
There are another action named "set speed 1" too, to reset it.
I'd bound right mouse button + wheel = speed change ; right mouse button + middle mouse button = reset speed; you can even combine mouse and keyboard, if you like.

**Comment by v-fox** [comment]
> It is not a feature many users want (i think)

Most users don't know what they want and we get things like "gnome-mpv" which has less functions than mpv's built-in scripted mini-menu >_< You have to endorse cool stuff to people for them to figure out effective approaches.

> if you like wheels, bound an hotkey...

That might work nicely , I'll try it out.
I've noticed that it's not possible to bind mouse combos: multiple mouse buttons, like "left+additional-left#1" or "right+middle" from your example. Is that possible to fix ? It also not clear how to delete a bind, the tooltip says «select a cell and hit "canc" to delete one», I figured it really is «highlight and hit "delete"». And after setting several binds whole binding menu stops reacting to presses until I completely restart Xt7. It also crashes if I hold "right arrow" key while it starts.

And is there a way to edit binding commands ? For example, I was trying to change "multiply speed" from 0.9 and 1.1 to 0.75 and 1.25 but it seems impossible. It's not an issue since "additional-left#1+wheel_up/down" works fine. In bomi there is a preference menu for such things called "control step". But to be fair, even it seems to unable to change those for binds.

**Comment by kokoko3k** [comment]
I'm sorry, but i still don't think the wheel would be an useful thing.
As per the ability to edit the action column, nope there is actually no way to add a custom action, nor editing an existing one; but it is a nice idea; maybe in a future release.

Mouse combos are complicated, because they'll need to introduce a delay on every other issued command, so if now i can left-click and have xt7 to pause almost immediately, with a combo it will have to wait a while to see if another button has been pressed to understand if the user want to make a combo or just pausing the stream (just an example)

Nice catch about the inability to bind multiple mouse buttons to an action (i forgot how implemented that years ago!)
Actually, xt7 is able to bind this scheme: 
[one Mouse button] + [a wheel direction] + [one key]
... where a key can be one printable character OR single/multiple modifiers

I'll change "canc" to "del" in my IT keyboard, my del button is named "canc" :)

**Comment by v-fox** [comment]
> I'm sorry, but i still don't think the wheel would be an useful thing.

Well, I, on the opposite, doubt the practicality of speed buttons. But either approach is not something "average user" desires. But then again, "average user" doesn't use advanced GUIs to console super-players.

> As per the ability to edit the action column, nope there is actually no way to add a custom action, nor editing an existing one; but it is a nice idea; maybe in a future release.

Cool !

> Mouse combos are complicated, because they'll need to introduce a delay on every other issued command, so if now i can left-click and have xt7 to pause almost immediately, with a combo it will have to wait a while to see if another button has been pressed to understand if the user want to make a combo or just pausing the stream (just an example)

That's true. And input delay is a horrendous thing.

> Nice catch about the inability to bind multiple mouse buttons to an action (i forgot how implemented that years ago!)

I've noticed thought that Xt7 does have delay on inputs. It actually would be a nice options to have:
- "instant mode" with no delay and combos based on button _un-presses_ (there is no way that a person could actually press 2 or more buttons completely instantaneously but instead inputs can be interpreted as "what were the last simultaneously held buttons ?"). That may break key "repeats" unless they are treated as fast "presses and un-presses" which they probably are.
- "delayed mode" with configurable delay and inputs interpreted as "what buttons were pressed from the start of the delay (any single button pressed / countdown started) to the end of it".

Personally, I would go for the first one.

> [one Mouse button] + [a wheel direction] + [one key]

Yeah, but that's too long, it would be easier to use just mouse or modifiers and those modifiers can be mapped on gamepad buttons too. Hopefully I have additional buttons on my mouse and DualShock isn't short on them either (while really I could use 4-6 more).

> I'll change "canc" to "del" in my IT keyboard, my del button is named "canc" :)

That's what it is, hah.

**Comment by kokoko3k** [comment]
Support for multiple simultaneous subtitle track is not included in the 0.12.381 release, but has just landed in the git repo:
https://github.com/kokoko3k/xt7-player-mpv/commit/373388b200d062510d9d39c918f9ae81a053dc47

**Comment by kokoko3k** [comment]
Opengl driver panel now has a little helper to add additional options:
https://github.com/kokoko3k/xt7-player-mpv/commit/9bc2a9816a6d04b6dc6f54786859ff59d9305405

**Comment by v-fox** [comment]
Seems Packman got the update.

> No button for mute/unmute; clicking hides/show the volume bar because some may like it, some may want to just use the wheel over the icon (on info are lost, because the actual volume is shown on the statusbar when you change it via scrollwheel).
> Maybe middleclicking on the button could mute/unmute (?)
> 
> -edit
> Yes: 7943119
> send me a beer :P

Is there something I could say to urge you to make an option to for putting mute/unmute and change zoom modes on left-click and hide & info on middle or right ? But that's low priority.
I would gladly sent you some beer, unfortunately I don't have enough rubbles for mail beer orders and in Russia any kind of alcohol mailing was forbidden until very recently. No company here does it yet.

> Fixed:
> 
> ```
> Relative text sizing for OSD and subtitles that scales with the size of the video output (it seems to scale accordingly to the height of the window (tested with vdpau,opengl and xv)
> ```

Mmm, it seems there is such mpv option. For some reason it barely noticeable with Xt7, probably because of video window proportions in non-fullscreen mode. I also found what seems like a bug - setting any kind of non-default "style" for any font makes it being "not found" by fontconfig and mpv is defaulting to Arial. Because of that I can't use my usual DejaVu Sans Bold.

> ```
> slick "playback information" (fixed the hotkey handling)
> ```

That's working great now !

> There is an option to scale the live preview size and to choose a different color for the seekbar.

I can't find scaling options and colour for seekbar seem to be ignored.
Also, colours for left-list are not automatically imported from theme and manually set "alternative background color" doesn't seem to get used (if it's purpose really is to provide alternating colours for odd and even filename rows).

I also just noticed that Xt7 seem to lack one more "Bomi function": the ability to "add" audio tracks. It's especially important if video and audio tracks are in separate files but those are not mkv-based.

I forgot to previously mention that 
- for some reason options "allow video panning via mouse-drag" and "Auto add eq2 and hue filters to video chain" are unchangeable for me;
- mouse pointer often "teleports" over the screen for no apparent reason;
- there is an annoying "double left click" bind for un/fullscreening, can it be removed ?

**Comment by v-fox** [comment]
> Actually, xt7 is able to bind this scheme:
> [one Mouse button] + [a wheel direction] + [one key]
> ... where a key can be one printable character OR single/multiple modifiers

I just tried binding Mouse4+Right_arrow (Additional mouse button №1 which is detected by Xt7 as "8") but it gets detected only as "Right_arrow" or "Mouse4" separately, depending which button is pressed first. I guess it's because it's not a "printable character OR modifier" but it would be quite handy to use as "L1/Left Button + right D-pad button" on my gamepad since "right D-pad button" is mapped to Right_arrow which is "seeking ahead 10 seconds".

While doing so, if I press "delete new" in hotkey dialogue, Xt7 hangs and if it gets killed, mpv runs without video window in background, waiting to be killed manually too.
Update: and it also consistently crashes if I press Ctrl+Shift which is "keyboard layout switch" with:

```
This application has raised an unexpected
error and must abort.

No keyboard event data.
HotkeyGrab.GetKey.122
```

**Comment by kokoko3k** [comment]
> Is there something I could say to urge you to make an option to for putting mute/unmute and change zoom modes on left-click and hide & info on middle or right ? But that's low priority.

FWIW, On kde, left clicking on the volume icon, pupups the volume slider.
middle clicking mute/unmutes it.
Why is it so important? It is just to memorize the right button (?)

> I would gladly sent you some beer

...of course i was joking ;)

> Mmm, it seems there is such mpv option. For some reason it barely noticeable with Xt7,

It actually works, to see it, just resize the video window vertically, and see that osd and subtitles DO scale.
The same applies to fullscreen, of course.

> setting any kind of non-default "style" for any font makes it being "not found" by fontconfig and mpv is defaulting to Arial.  Because of that I can't use my usual DejaVu Sans Bold.

mpv supports settings of bold fonts, no other styles are supported.
So it is expected that italic,condensed,light and so on will not work.
On the other hand, bold does, and it works for me (the osd in the following screenshot is dejavu sans, while the subtitle is dejavu sans bold):
![xt7bold jpg](https://cloud.githubusercontent.com/assets/1230726/11182834/16e8eda4-8c6e-11e5-9eeb-4d64105bf430.jpg)

> I can't find scaling options and colour for seekbar seem to be ignored.

seekbar color works for me:
![xt7-progressbar-color](https://cloud.githubusercontent.com/assets/1230726/11182894/a18c5752-8c6e-11e5-857e-afe0ed7e7b14.jpg)

scaling option is here:
![xt7-scaling](https://cloud.githubusercontent.com/assets/1230726/11184053/aa233370-8c75-11e5-813d-4d53bda00710.jpg)

> Also, colours for left-list are not automatically imported from theme and manually set "alternative background color" doesn't seem to get used (if it's purpose really is to provide alternating colours for odd and even filename rows).

The filesystem browser does not support alternate row colors because it is a builtin gambas component.
You could try to ask gambas devs to implement such feature.
Theme colors managing was wrong, i'll try to fix them in an upcoming release.
Btw i just fixed them a bit; if you hit the reset button, now it should change the color used by xt7 accordingly to the one used in the theme.
The problem is that if youchange the theme palette, xt7 will still use the old one, requireing you to click the reset button again; this needs to be fixed.

> I also just noticed that Xt7 seem to lack one more "Bomi function": the ability to "add" audio tracks. It's especially important if video and audio tracks are in separate files but those are not mkv-based.

while i don't think it would be a widely used function, i'll plan to add it.

> for some reason options "allow video panning via mouse-drag" and "Auto add eq2 and hue filters to video chain" are unchangeable for me;

Expected. they are two settings from the old mplayer code; i'll left them disabled for everyone till i find a way to implement them, you won't say that, but is not simple.
Of course i can hide the controls completely, but i need them to remember me to think of it.

> mouse pointer often "teleports" over the screen for no apparent reason;

I definitely need more info on that!

> there is an annoying "double left click" bind for un/fullscreening, can it be removed ?

Well, not now, need to add another option to allow that; will do in the hotkeys window.

> I just tried binding Mouse4+Right_arrow  but it gets detected only as "Right_arrow" or "Mouse4" separately

..well, is not that it gets detected "separately", it detects Mouse0 if you press a key.
It is a bug, i'll look into it.

> if I press "delete new" in hotkey dialogue, Xt7 hangs and if it gets killed, 

Not happening here, could you reproduce?

> mpv runs without video window in background, waiting to be killed manually too.

mpv should be killed in a matter of 10 seconds by a leftover xt7 process, did you killed it first?

> Update: and it also consistently crashes if I press Ctrl+Shift which is "keyboard layout switch" with [..]

It is not happening here, but i can handle the error; thanks for attaching the debug infos.
The error handling is fixed in the latest git; handling the hotkey 

Two side notes:
This discussion is becoming difficult to follow ; could you please open separate issues for each of your findings?
Having you to use git code instead of the packman releases would be better, thanks.

**Comment by kokoko3k** [comment]
A little summary of the latest commits that may interest you follows:
- Allow to add external audio files to the current video track.
- Allow to disable doubleclick to go fullscreen
- Rewrote hotkeys handling routines
- Use theme palette when resetting colors in the style tab
- New helper to choose additional opengl video driver options
-  Support for secondary subtitle track
- "canc" key is actually the "del" key. 
- Allow to mute/unmute by middle-clicking over the volume icon
- Hotkey to show runtime infos acts now as a switch, no timeouts.
- Do not move video frame when showing main toolbar in fullscreen
- Allow to Change the live preview size and the seekbar color
- Restore maximized state under openbox too 
- Field for setting the background image under the panel was missing.

Now i am closing this issue one because, it is difficult to follow aggregate multiple issues.
If you find something new, or just want to talk about a specific issue, please open a report for each one.
Since those commits interest you and probably they need some deep testing, i encourage you again to build xt7 from git to test them, (build process takes a bunch seconds even on slower machines).

**Comment by kokoko3k** [comment]
Just to let you know, now xt7 is using the following for stats:
https://github.com/Argon-/mpv-stats
